lib/AST/ASTScopeLookup.cpp

```diff
@@ -675,3 +675,31 @@ std::pair<CaseStmt *, CaseStmt *> ASTScopeImpl::lookupFallthroughSourceAndDest(
 
   return { nullptr, nullptr };
 }
+
+bool ASTScopeImpl::isInMacroArgument(SourceFile *sourceFile,
+                                     SourceLoc loc) {
+  if (!sourceFile || sourceFile->Kind == SourceFileKind::Interface)
+    return false;
+
+  if (loc.isInvalid())
+    return false;
+
+  auto *fileScope = sourceFile->getScope().impl;
+  auto *scope = fileScope->findInnermostEnclosingScope(loc, nullptr);
+  do {
+    if (scope->isMacroArgumentScope())
+      return true;
+
+    // If we've reached a source file scope, we can't be inside of
+    // a macro argument. Either this is a top-level source file, or
+    // it's macro expansion buffer. We have to check for this because
+    // macro expansion buffers for freestanding macros are children of
+    // MacroExpansionDeclScope, and child scopes of freestanding macros
+    // are otherwise inside the macro argument.
+    if (scope->getClassName() == "ASTSourceFileScope")
```

Comment:
Is there a better way to check the `ASTScope` kind?

Reply:
No, and it looks like you're not the first to lament that this is the only way to check:

```
  // FIXME: Proper isa/dyn_cast support would be better than a string
  // comparison here.
  while (parent->getClassName() == "GenericParamScope")
    parent = parent->getLookupParent().get();
```